Updated transformations API guide (#3386)
openvinotoolkit/openvino: docs/nGraph_DG/nGraphTransformation.md

and is located in the `inference-engine/src/transformations` directory.

Transformations root directory contains two folders:
* `ngraph_ops` - Contains internal opset operations that are common for plugins.
* `transformations` - Includes all transformations, utils, runtime info attributes, and pass managers.

> All internal operations and transformations located inside [Transformation Library](group__ie__transformation__api.html) are allowed for usage inside plugins.

Reviewer: ```suggestion
All internal operations and transformations located inside the [Transformation Library](group__ie__transformation__api.html) can be used inside plugins.
```

Author: done.

> All legacy operations and transformations were moved to legacy library and not suggested to use.

Reviewer: ```suggestion
All legacy operations and transformations were moved to a legacy library and are not recommended to be used.
```

Author: done.

To decide where to store your transformation code, please follow these rules:

1. If it is a plugin-specific transformation and cannot be reused by other plugins, keep source code inside plugin.

> 2. If this transformation relates to opset operation conversion or optimization then keep sources inside the transformation library.

Reviewer: ```suggestion
2. If this transformation relates to opset operation conversion or optimization, keep sources inside the transformation library.
```

Author: done

> After you decide where to store your transformation code, you can start developing your own nGraph transformation.

Reviewer 2: Could you update a paragraph about ngraph function and add that nGraph function contains pointers to  `Parameter`, `Result` and `Sink`?

Author: Actually I want to divide this doc into several parts. In this PR I juts update transformations API related topics.

So let's go through each of these steps.

### Create a pattern
Pattern is a single root `ngraph::Function`. But the only difference is that you do not need to create a function object, you just need to create and connect opset or special pattern operations.
Then you need to take the last created operation and put it as a root of the pattern. This root node will be used as a root node in pattern matching.
> **NOTE**: Any nodes in a pattern that have no consumers and are not registered as root will not be used in pattern matching. 

@snippet example_ngraph_utils.cpp pattern:simple_example

The `Parameter` operation in the example above has type and shape specified. These attributes are needed only to create Parameter operation class and will not be used in pattern matching.

> For more pattern examples please follow the [pattern matching](#pattern_matching) section.

Reviewer: ```suggestion
For more pattern examples, refer to the [pattern matching](#pattern_matching) section.
```

Author: done.

![graph_rewrite_efficient_search] 

> > **NOTE**: GraphRewrite execution algorithm can not set manually and depends only on root nodes registered inside MatcherPasses.

Reviewer: ```suggestion
> **NOTE**: GraphRewrite execution algorithm cannot be set manually and depends only on root nodes registered inside MatcherPasses.
```

Author: done

## Disabling/Enabling specific transformations for plugin X	 <a name="disabling_transformation"></a>

> In transformation library we offer to plugins transformations like CommonOptimizations that contains predefined sequence of transformations.

Reviewer: ```suggestion
In transformation library, we provide plugins transformations like CommonOptimizations, which contains predefined sequence of transformations.
```

Author: done.

> So we also provide a tool that helps to disable or partially disable particular transformations in transformation pipeline.

Reviewer: ```suggestion
We also provide a tool that helps to disable or partially disable particular transformations in a transformation pipeline.
```

Author: done.

> For example, if plugin uses CommonOptimization transformation and want to disable ConvertGELU transformation then inside plugin we have to take PassConfig instance

Reviewer: ```suggestion
For example, if a plugin uses the CommonOptimization transformation and needs to disable the ConvertGELU transformation, then inside the plugin we have to take the PassConfig instance
```

Author: done

from pass::Manger and call disable method.

@snippet example_ngraph_utils.cpp ngraph:disable_gelu

> In some cases we want to disable transformation for some condition:

Reviewer: ```suggestion
In some cases, we need to disable transformation for some condition:
```

Author: done.

@snippet example_ngraph_utils.cpp ngraph:disable_callback

> And in some cases pass::Manager pipelines inside transformations may have transformations disabled by default but enabled insde plugins.

Reviewer: ```suggestion
In some cases, pass::Manager pipelines inside transformations may have transformations disabled by default but enabled inside plugins.
```

Author: done.

@snippet example_ngraph_utils.cpp ngraph:disabled_by_default

> PassConfig instance taken from pass::Manager is shared across all registered transformations including nested transformations. So it doesn't matter where we work with this object (before passes registration or after).

Reviewer: ```suggestion
PassConfig instance taken from pass::Manager is shared across all registered transformations including nested transformations. So it does not matter where we work with this object (before passes registration or after).
```

Author: done.
